lib/bridge/MatrixHandler.js

```diff
@@ -770,6 +769,83 @@ MatrixHandler.prototype._onMessage = Promise.coroutine(function*(req, event) {
     yield Promise.all(promises);
 });
 
+MatrixHandler.prototype._sendIrcAction = Promise.coroutine(
+    function*(req, ircRoom, ircClient, ircAction, event) {
+
+    // Send the action as is if it is not a text message 
+    //  Also, check for the existance of the getSplitMessages method.
+    if (event.content.msgtype !== "m.text" ||
+        !(ircClient.unsafeClient && ircClient.unsafeClient.getSplitMessages)) {
+        yield this.ircBridge.sendIrcAction(ircRoom, ircClient, ircAction);
+        return;
+    }
+
+    let text = event.content.body;
+
+    // Generate an array of individual messages that would be sent
+    let potentialMessages = ircClient.unsafeClient.getSplitMessages(ircRoom.channel, text);
+    let lineLimit = ircRoom.server.getLineLimit();
+
+    if (potentialMessages.length <= lineLimit) {
+        yield this.ircBridge.sendIrcAction(ircRoom, ircClient, ircAction);
+        return;
+    }
+
+    // Message body too long, upload to HS instead
+    try {
+        // Use the current time as the name of the file
+        //  strip off day of week and timezone
+        //  replace whitespace with underscores
+        // result e.g : Aug_01_2016_13:41:44.txt
+        let fileName = new Date().toString()
+            .match(/.*?\s(.*\:[0-9]*\:[0-9]*).*/)[1]
+            .replace(/\s/g,'_') + ".txt";
+
+        // Try to upload as a file and get URI
+        //  (this could fail, see the catch statement)
+        let response = yield this.ircBridge.sendMatrixPlaintextFile(fileName, text);
+        let result = JSON.parse(response);
+
+        // Alter event object so that it is treated as if a file has been uploaded
+        event.content.url = result.content_uri;
+        event.content.msgtype = "m.file";
+        event.content.body = fileName;
+
+        // Recreate action from modified event
+        let bigFileIrcAction = IrcAction.fromMatrixAction(
+            MatrixAction.fromEvent(
+                this.ircBridge.getAppServiceBridge().getClientFactory().getClientAs(), event
+            )
+        );
+
+        // Send the recreated action in place of ircAction
+        yield this.ircBridge.sendIrcAction(ircRoom, ircClient, bigFileIrcAction);
+    }
```

Comment:
After thinking about this, I think this try/catch is too general and should be made more precise. There's a few reasons for this:
 - The `catch` block doesn't know how far along the `try` got to, whether it uploaded the text file, read the response, or sent the IRC action, so it can't reasonably retry any bits without doing the lot again.
 - Because there is so much in the `try` block, it need to do weird subtle things like `:831` in the `catch` (where if the URL property is set on the `event` but sending the IRC action fails means you then need to back out the changes you made to the event).
 - The `catch` block is assuming failure is due to network errors (as shown by the log line), but this may not be true.